video/out/gpu/context.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include "video/out/vo.h"
+
+#include "config.h"
+#include "ra.h"
+
+enum ra_api_type {
+    RA_API_ANY    = 0,
+    RA_API_OPENGL,
+};
+
+struct ra_ctx_opts {
+    int allow_sw;      // allow software renderers
+    int want_alpha;    // create an alpha framebuffer if possible
+    int debug;         // enable debugging layers/callbacks etc.
+    bool probing;      // the backend was auto-probed
+    int swchain_depth; // max number of images to render ahead
+};
+
+struct ra_ctx {
+    struct vo *vo;
+    struct ra *ra;
+    struct mpv_global *global;
+    struct mp_log *log;
+
+    struct ra_ctx_opts opts;
+    const struct ra_ctx_fns *fns;
+
+    bool flip_v; // flip the rendered image vertically (set by the ra_ctx)
+    void *priv;
+
+#if HAVE_GL
+    struct ra_gl_ctx *gl_priv;
+#endif
```

Comment:
No ifdef needed. But why is this field even there.

Reply:
For the same reason as why `struct vo` has the x11 etc. in there -> it's basically so we can subtype `struct ra_ctx`. Basically, the `ra_gl_ctx_*` helpers use this internally, and I consider this saner than trying to require that `void *priv` can be aliased to `struct ra_gl_ctx *` or something.

Reply:
Well, it should be handled in a different way.

vo does it only because it was the way MPlayer did it. (And you could see the results of that, until mpv slowly moved to a more disciplined approach.)

Reply:
How would you handle it? The next-best solution would be to add `ra_gl_ctx *p` to each context's `struct priv` and write a bunch of 2-line wrappers that look like this:

```c
static T do_thing(struct ra_ctx *ctx)
{
    struct priv *p = ctx->priv;
    ra_gl_ctx_do_thing(p->gl_priv);
}

...
    .do_thing = do_thing
```

That would require in a lot of boilerplate, and I'm not a huge fan of writing it all.

Reply:
I'd just have separated ra_context into several orthogonal components.

Reply:
See earlier comment. That's not something we can really do, or I would have done it, too.

Reply:
FWIW, I'm concerned about the case where a GL backend only wants to override some of the generic GL methods. For example, angle and dxinterop use the generic GL implementation of .depth(): ra_gl_ctx_depth, which depends on gl_priv, but at some point I want to replace this with a custom implementation, since I think the generic implementation returns the wrong value.

Reply:
Yes, that was very much the intent of adding these fields to the `ra_ctx_fns`: so contexts can override them based on per-platform requirements if they have a need to do so. (e.g. perhaps because the GL stuff returns broken values, or because there's a “better” way).

That said, I think it's orthogonal to this. The entry-point wouldn't go away from `ra_ctx_fns` in this proposed refactor, it would just force you to write a custom wrapper function that redirects it to `ctx->priv->gl_priv` in cases where you want to re-use the `ra_gl_ctx` helper.